### ISSUE TITLE
Masks in cross attention

### ISSUE THREAD

**Issue by gpantaz** [lucidrains/x-transformers#183]
Hello, thank you for this fantastic repo. It is literally amazing!

I would like to ask you a question regarding the masking in cross attention. In particular, I would like to implement custom cross attention blocks in between the layers of a language model (similar to flamingo):

![image](https://github.com/lucidrains/x-transformers/assets/30076767/0ad7b88b-7809-46dc-9e7a-df37205305b6)

I started with only the `Attention` module without any additional FF blocks so that I can implement the gating myself. From my understanding:
1. The input corresponds to the language embeddings (will be padded during training)
2. The context corresponds to the vision embeddings (fixed for every item in the batch)

Looking at the forward pass of the `Attention` module it accepts 3 different masks `context_mask`, `mask`, & `attn_mask`

From line 792 I understand that the `mask` is replaced by `context_mask` and thus not required.
https://github.com/lucidrains/x-transformers/blob/7395ebd9bfaea85ef8358e1d46ca176351058017/x_transformers/x_transformers.py#L792

For the `context_mask` in my case I assume that a full mask is enough since the vision embeddings are not padded.

Q1: What is the role of the `attn_mask`? I assume that the padded tokens from the input will cross-attend to the image tokens but that is okay since the input tokens will not attend to themselves correct?

Q2: The input has a causal structure - do I need to provide a causal mask? I assume that this is not required again because the input tokens do not attend to themselves correct?

If my assumptions are correct would this be sufficient as an intermediate cross-attention block?

```
from x_transformers.x_transformers import Attention

attn = Attention(512, heads=8, dropout=0.1, flash=False)

input_embeds = torch.randn(2, 10, 512)

# Actually not required since the `mask` is replaced by the `context_mask`
# padding_mask = torch.zeros(input_embeds.shape[0], input_embeds.shape[1])
# padding_mask[0, :6] = 1
# padding_mask[1, :] = 1

# Actually not required since input embeds do not attend to themselves (Q2)
# attn_mask = _make_causal_mask(
#     input_embeds.shape[:-1], dtype=input_embeds.dtype, device=input_embeds.device
# )

context = torch.randn(2, 20, 512)
context_mask = torch.ones(context.shape[0], context.shape[1]).bool()

out = attn(
    input_embeds,
    context=context,
    context_mask=context_mask,
)[0]
```

Thank you!

**Comment by lucidrains** [comment]
@gpantaz hey George, thanks for the kind words

your example is correct usage! `mask` and `context_mask` are the same for self and cross attention respectively. both refer to masking out the key / values, commonly known as `key_padding_mask` - so for attention matrix `(i, j)`, it would be masking out the `j` dimension

`attn_mask` refers to masking of the entire attention matrix, so would be of shape `(i, j)` and would denote which language token can attend to which vision token. this gives you full control over what attends to what

**Comment by gpantaz** [comment]
Thanks for the clarification!
